airflow/providers/amazon/aws/hooks/s3.py

```diff
@@ -292,18 +302,24 @@ def list_keys(
             'PageSize': page_size,
             'MaxItems': max_items,
         }
-
         paginator = self.get_conn().get_paginator('list_objects_v2')
         response = paginator.paginate(
-            Bucket=bucket_name, Prefix=prefix, Delimiter=delimiter, PaginationConfig=config
+            Bucket=bucket_name,
+            Prefix=prefix,
+            Delimiter=delimiter,
+            PaginationConfig=config,
+            StartAfter=start_after_key,
+        )
+        # JMESPath to query directly on paginated results
+        filtered_response = response.search(
+            "Contents[?to_string("
+            "LastModified)<='\"{}\"' && "
+            "to_string(LastModified)>='\"{"
```

Comment:
does not handle case where you have exactly _one of_ `to_datetime` and `start_after_datetime` (i.e. not both)

also pendulum and datetime are formatted differently. observe:
```python
>>> import pendulum
>>> '{}'.format(pendulum.now())
'2021-07-06T21:22:11.514220-07:00'
>>> from datetime import datetime
>>> '{}'.format(datetime.now())
'2021-07-06 21:22:35.418199'
```

if you compare the objects directly, your feature could support both pendulum _(which is used commonly with airflow)_ and datetime; but by using string formatting you really cannot use both.

Reply:
@potiuk  @dstandish @pateash I have pushed the changes just now.

Created `ResponseFilter` class to parse and filter the s3 boto3 response based on various operations defined by the user.  This class can be extended as required for parsing and filtering the response based on different object filters operation. For now, I have added the following allowed operations.
```
 allowed_operations = {
        "lte": operator.le,
        "gte": operator.ge,
        "gt": operator.gt,
        "lt": operator.lt,
        "eq": operator.eq,
    }
```

`ResponseFilter.filter(object_filter=object_filter)`  puts the filtering in the user's hands (through the use of a filter and would remove the ambiguity of strictly greater than vs greater than (and same with less than). Please refer to the commits in this [thread](https://github.com/apache/airflow/pull/16821/commits/4ef750ea486f76c98f7e5927b12fc62b4767913c) and [test changes](https://github.com/apache/airflow/pull/16821/commits/a7468c0d1591c83fd1a1c7f3ca8e32a2750e4f5b).

I have added the comment for the implementation details in the code.

Reply:
i'm not sure i like the ResponsFilter approach.

if you want to use the `object_filter` approach, why not just let the user put an arbitrary callable that does the filtering that the user needs.  e.g. `lambda x: x['LastModified'] > some_val`. or `lambda x: other_val > x['LastModified'] > some_val`

to me this seems much simpler to implement (cus it removes the entire ResponseFilter` class) and quite simple and transparent to use.

i am not sure we need to add this abstraction layer -- that's just another thing to have to learn.

alternatively, i would be fine doing `from_datetime` and `to_datetime` where from is evaluated `>=` and to is evaluated `<`

what do yall think?

Reply:
Sure @dstandish, I think the `to` and `from` approach keeps the code simple and doesn't add additional complexity to it. `ResponseFilter` in my opinion makes things more generic for any filter operations in the future but it does add additional complexity to it. Would be happy to accommodate the change once others comment and conclude on the approach.

@potiuk what do you think?

Reply:
@dstandish I thought about it, if we take `ResponseFilter` route, it will add additional complexity to the code and people might need to understand the approach. Hence for simplicity of the codebase, I am going forward with `from` and `to` approach. I would be renaming it to `lastmodified_datetime_gte` (>=) and `lastmodified_datetime_lt` (<) .